### PR TITLE
AIRFLOW-45: Support Hidden Airflow Variables

### PR DESCRIPTION
**Pull request by mattuuh7**
Dear Airflow Maintainers,

Please accept this PR that addresses the following issues:
- https://issues.apache.org/jira/browse/AIRFLOW-45

Reminders for contributors:
- Your PR's title must reference an issue on [Airflow's JIRA](https://issues.apache.org/jira/browse/AIRFLOW/). For example, a PR called "[AIRFLOW-1] My Amazing PR" would close JIRA issue #1. Please open a new issue if required!
- You must add an [Apache License header](http://www.apache.org/legal/src-headers.html) to all new files.
- Please squash your commits when possible and follow the [7 rules of good Git commits](http://chris.beams.io/posts/git-commit/#seven-rules).

### CODE REVIEW

File: airflow/configuration.py [hunk 1]
```diff
@@ -372,6 +375,9 @@ def run_command(command):
 # default_principal = admin
 # default_secret = admin
 
+[admin]
+hide_confidential_ui_fields = true
```
Comment (criccomini) [inline airflow/configuration.py]: Should be `True` just to be more Python-esque, e.g.

```
 # statsd_on =  False
```

File: airflow/www/views.py [hunk 2]
```diff
@@ -2005,15 +2020,19 @@ class DagPickleView(SuperUserMixin, ModelView):
 admin.add_view(mv)
 '''
 
-
 class VariableView(wwwutils.LoginMixin, AirflowModelView):
+    def hidden_field_formatter(view, context, model, name):
```
Comment (criccomini) [inline airflow/www/views.py]: Nit: recommend dropping below the variable declarations.

File: airflow/configuration.py (hunk 1 above)
Comment (r39132) [inline airflow/configuration.py]: So, why True as default and not False as default? False preserves backward compatibility with respect to UI experience.

Comment (criccomini) [inline airflow/configuration.py]: The issue is if we default to false, then passwords are visible by default.

**Comment by criccomini** [comment]
Wondering whether we want to include extras in the filtering. It's annoying, and I don't think that there is any secret stuff in there for existing connections. Thoughts?

**Comment by criccomini** [comment]
Note: installed locally, and this did appear to work.

**Comment by criccomini** [comment]
Actually, I take that back. It worked for the hooks. For variables, it only works for very specific fields:

```
    test    test    
    password    ********    
    secret_password test    
```

I would have expected either all to be encrypted, or anything with 'password'/'secret' in it.

File: airflow/configuration.py [hunk 3]
```diff
@@ -142,7 +142,10 @@ def run_command(command):
     },
     'github_enterprise': {
         'api_rev': 'v3'
-    }
+    },
+    'admin': {
+        'hide_confidential_ui_fields': True,
+    },
 }
```
Comment (r39132) [inline airflow/configuration.py]: Make sure you have specified the config value everywhere it is needed. I recall there are at least 3 places where new config needs to be specified.

For example, look at `dags_are_paused_at_creation` :

```
(venv) Sid-As-MacBook-Pro:airflow siddharth$ grep -ri dags_are_paused_at_creation . | grep -v Binary | grep -v static
./configuration.py:        'dags_are_paused_at_creation': True,
./configuration.py:dags_are_paused_at_creation = True
./configuration.py:dags_are_paused_at_creation = False
./models.py:    is_paused_at_creation = configuration.getboolean('core', 'dags_are_paused_at_creation')
```

Comment (mattuuh7) [inline airflow/configuration.py]: @criccomini updated the logic, it should check if any of the `key_name` contains any words in the confidential_fields list that models after [sentry's list](https://github.com/getsentry/sentry/blob/master/src/sentry/constants.py#L158)

screen shot 2016-05-20 at 3 57 19 pm

**Comment by r39132** [comment]
Thx for providing a screen shot.. generally a requirement for all UI-changing PRs, though you can attach the screen shot here or in the JIRA. Both work.

**Comment by criccomini** [comment]
@mattuuh7, this is looking good.
1. Can you remove extras from being masked? I think the cons outweigh the pros of masking it. (I know this is annoying, considering how much time you spent trying to figure out how to mask it. Sorry. :()
2. Can you rename hide_encrypted_ui_fields to hide_sensitive_variable_fields?
3. Could you add some docs either to the .rst or to the config comments saying which variables get masked (password, secret, etc).
4. Could you squash your commit, and rename both the commit message and title of the PR as: 'AIRFLOW-45: Support Hidden Airflow Variables'

**Comment by criccomini** [comment]
When running this off of master, I get:

```
  File "/Users/chrisr/Code/airflow/airflow/www/views.py", line 2038, in hidden_field_formatter
    return getattr(model, name)
  File "build/bdist.macosx-10.10-intel/egg/sqlalchemy/orm/attributes.py", line 293, in __get__
    return self.descriptor.__get__(instance, owner)
  File "/Users/chrisr/Code/airflow/airflow/models.py", line 3200, in get_val
    "Can't decrypt _val, configuration is missing")
AirflowException: Can't decrypt _val, configuration is missing
```

This can be reproduced by going to a fresh airflow install, running `airflow initdb`, `airflow webserver`, and opening the variables UI.

**Comment by criccomini** [comment]
Removing `val` from:

```
column_list = ('key', 'val', 'is_encrypted',)
```

Made the exception go away. The patch appears to work without it, but `val` isn't shown in the table view anymore.

**Comment by criccomini** [comment]
I've determined the issue was with some rows that somehow persisted even after I ran `airflow initdb`. Running `airflow resetdb` fixed the issue.

**Comment by criccomini** [comment]
Ready to merge. Could you change the commit message to:

AIRFLOW-45: Support Hidden Airflow Variables

**Comment by mattuuh7** [comment]
@criccomini done with the change. please review again. thx

**Comment by mattuuh7** [comment]
thanks @criccomini

**Pull request merged**